sql/sql_repl.cc

```diff
@@ -3369,6 +3369,7 @@ static my_bool kill_callback(THD *thd, kill_callback_arg *arg)
       thd->variables.server_id == arg->slave_server_id)
   {
     arg->thd= thd;
+    if (WSREP(thd)) mysql_mutex_lock(&thd->LOCK_thd_data);
```

Comment:
Can you explain why taking this lock is needed?

Reply:
Repeating same test calls kill_zombie_dump_threads() which calls awake_no_mutex() that expects that lock is set on thd->LOCK_thd_data.
Prior calling awake_no_mutex(), kill_callback is used to find thread which matches slave id thread.
This is where needed lock needs to be set.